### PR TITLE
chore: lazy load imports to speed up load time [MLG-455]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description
loading harness takes around 0.75s on my machine. there's quite a few imports that can be lazily-imported to reduce load time.

**Before**
```
Program: determined.cli.cli

0.753 <module>  harness/determined/__init__.py:1
└─ 0.725 <module>  harness/determined/core/__init__.py:1
   ├─ 0.674 <module>  harness/determined/core/_checkpoint.py:1
   │  └─ 0.673 <module>  harness/determined/tensorboard/__init__.py:1
   │     └─ 0.660 <module>  harness/determined/tensorboard/metric_writers/__init__.py:1
   │        └─ 0.659 <module>  harness/determined/tensorboard/metric_writers/callback.py:1
   │           └─ 0.658 <module>  harness/determined/util.py:1
   │              └─ 0.655 <module>  harness/determined/common/__init__.py:1
   │                 ├─ 0.378 <module>  harness/determined/common/api/__init__.py:1
   │                 │  ├─ 0.199 <module>  harness/determined/common/api/authentication.py:1
   │                 │  │  ├─ 0.152 <module>  harness/determined/common/api/bindings.py:2
   │                 │  │  │  └─ 0.124 <module>  requests/__init__.py:39
   │                 │  │  │        [224 frames hidden]  requests, charset_normalizer, <built-...
   │                 │  │  └─ 0.021 <module>  filelock/__init__.py:7
   │                 │  │        [20 frames hidden]  filelock, <built-in>
   │                 │  ├─ 0.131 <module>  harness/determined/common/api/request.py:1
   │                 │  │  └─ 0.126 <module>  lomond/__init__.py:1
   │                 │  │        [79 frames hidden]  lomond, platform, os, subprocess, <bu...
   │                 │  └─ 0.034 <module>  harness/determined/common/api/profiler.py:1
   │                 │     └─ 0.026 decorate  backoff/_decorator.py:152
   │                 │           [35 frames hidden]  backoff, asyncio, <built-in>, concurrent
   │                 ├─ 0.137 <module>  harness/determined/common/util.py:1
   │                 │  └─ 0.125 <module>  urllib3/__init__.py:3
   │                 │        [266 frames hidden]  urllib3, re, sre_compile, sre_parse, ...
   │                 ├─ 0.061 <module>  ruamel/yaml/__init__.py:1
   │                 │     [67 frames hidden]  ruamel, base64, struct, <built-in>
   │                 ├─ 0.051 <module>  harness/determined/common/storage/__init__.py:1
   │                 │  └─ 0.041 <module>  harness/determined/common/storage/hdfs.py:1
   │                 │     └─ 0.037 <module>  hdfs/__init__.py:4
   │                 │           [91 frames hidden]  hdfs, multiprocessing, pickle, <built...
   │                 └─ 0.018 <module>  harness/determined/common/context.py:1
   │                    └─ 0.013 <module>  pathspec/__init__.py:30
   │                          [32 frames hidden]  pathspec, <built-in>, typing
   └─ 0.040 <module>  harness/determined/core/_distributed.py:1
      └─ 0.032 <module>  harness/determined/ipc.py:1
         └─ 0.030 <module>  psutil/__init__.py:21
               [41 frames hidden]  psutil, <built-in>, enum, collections...
```

After
```
Program: determined.cli.cli

0.487 <module>  harness/determined/__init__.py:1
├─ 0.433 <module>  harness/determined/core/__init__.py:1
│  ├─ 0.388 <module>  harness/determined/core/_distributed.py:1
│  │  ├─ 0.374 <module>  harness/determined/util.py:1
│  │  │  └─ 0.369 <module>  harness/determined/common/__init__.py:1
│  │  │     ├─ 0.248 <module>  harness/determined/common/api/__init__.py:1
│  │  │     │  ├─ 0.226 <module>  harness/determined/common/api/authentication.py:1
│  │  │     │  │  ├─ 0.136 compile  None
│  │  │     │  │  │     [2 frames hidden]  <built-in>
│  │  │     │  │  └─ 0.078 <module>  harness/determined/common/api/bindings.py:2
│  │  │     │  │     └─ 0.061 <module>  requests/__init__.py:39
│  │  │     │  │           [156 frames hidden]  requests, http, re, sre_compile, sre_...
│  │  │     │  └─ 0.009 compile  None
│  │  │     │        [2 frames hidden]  <built-in>
│  │  │     ├─ 0.071 <module>  harness/determined/common/util.py:1
│  │  │     │  └─ 0.063 <module>  urllib3/__init__.py:3
│  │  │     │        [187 frames hidden]  urllib3, re, sre_compile, sre_parse, ...
│  │  │     ├─ 0.018 <module>  ruamel/yaml/__init__.py:1
│  │  │     │     [33 frames hidden]  ruamel, base64, <built-in>, struct, abc
│  │  │     └─ 0.016 <module>  harness/determined/common/storage/__init__.py:1
│  │  └─ 0.009 compile  None
│  │        [2 frames hidden]  <built-in>
│  ├─ 0.021 <module>  harness/determined/core/_checkpoint.py:1
│  │  └─ 0.019 <module>  harness/determined/tensorboard/__init__.py:1
│  └─ 0.016 compile  None
│        [2 frames hidden]  <built-in>
├─ 0.029 <module>  harness/determined/_trial_controller.py:1
│  ├─ 0.017 <module>  harness/determined/profiler.py:1
│  │  └─ 0.012 <module>  psutil/__init__.py:21
│  │        [22 frames hidden]  psutil, <built-in>, enum, collections
│  └─ 0.010 compile  None
│        [2 frames hidden]  <built-in>
├─ 0.012 compile  None
│     [2 frames hidden]  <built-in>
└─ 0.009 <module>  harness/determined/_info.py:1

```
## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/common/api/profiler.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
+import time
 from typing import Any, Dict, List, Optional
 
-import backoff
```
Comment (rb-determined-ai) [inline harness/determined/common/api/profiler.py:3]: why not remove the other two instances of `@backoff` and just drop the dependency altogether?

**Review by rb-determined-ai — COMMENTED** [review]
add tests?  See `harness/tests/test_imports.py`

File: harness/determined/common/api/profiler.py [hunk 2]
```diff
@@ -41,11 +35,23 @@ def post_trial_profiler_metrics_batches(
...
-    api.post(
-        master_url,
-        "/api/v1/trials/profiler/metrics",
-        json={"batches": [b.__dict__ for b in batches]},
-    )
+    backoff_interval = 1
+    max_tries = 2
+    tries = 0
+
+    while tries < max_tries:
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/common/api/profiler.py:42]: 🤦

File: harness/tests/test_imports.py [hunk 3]
```diff
@@ -37,6 +37,10 @@ def test_import_side_effects() -> None:
             determined.cli.cli.main()
 
             bad = {
+                "^backoff\\.*",
```
Comment (rb-determined-ai) [inline harness/tests/test_imports.py:40]: not necessary if it's not a dependency

File: master/static/srv/enrich_task_logs.py [hunk 4]
```diff
@@ -119,17 +118,22 @@ def run(self) -> None:
             # Timeout met.
             self.ship()
 
-    @backoff.on_exception(  # type: ignore
-        lambda: backoff.full_jitter(SHIPPER_FAILURE_BACKOFF_SECONDS),
```
Comment (rb-determined-ai) [inline master/static/srv/enrich_task_logs.py:123]: what you implemented was a constant backoff, but what you are deleting [is a lot more involved](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/)

Comment (rb-determined-ai) [reply master/static/srv/enrich_task_logs.py:123]: well, not that much more involved.  Looks like the full jitter algorithm is just:

```python
sleep = random_between(0, min(cap, base * 2**attempt))
```

Comment (azhou-determined) [reply master/static/srv/enrich_task_logs.py:123]: so, i looked into the backoff source code trying to figure out how to reimplement full_jitter, and it turns out we weren't implementing it correctly anyway. as is, this code does constant (but random) backoff. i consulted Jerry/Bradley and they said constant is fine for this method.

**Review by rb-determined-ai — APPROVED** [review]
awesome, I love speedups.

**Pull request merged**